packages/runtimes/hmr/src/loaders/hmr-runtime.js

```diff
@@ -300,10 +315,73 @@ function hmrApply(bundle /*: ParcelRequire */, asset /*:  HMRAsset */) {
   }
 }
 
+function hmrDelete(bundle, id) {
+  let modules = bundle.modules;
+  if (!modules) {
+    return;
+  }
+
+  if (modules[id]) {
+    // Collect dependencies that will become orphaned when this module is deleted.
+    let deps = modules[id][1];
+    let orphans = [];
+    for (let dep in deps) {
+      let parents = getParents(module.bundle.root, deps[dep]);
+      if (parents.length === 1) {
+        orphans.push(deps[dep]);
+      }
+    }
+
+    // Delete the module. This must be done before deleting dependencies in case of circular dependencies.
+    delete modules[id];
+    delete bundle.cache[id];
+
+    // Now delete the orphans.
+    orphans.forEach(id => {
+      hmrDelete(module.bundle.root, id);
+    });
+  } else if (bundle.parent) {
+    hmrDelete(bundle.parent, id);
+  }
+}
+
 function hmrAcceptCheck(
   bundle /*: ParcelRequire */,
   id /*: string */,
   depsByBundle /*: ?{ [string]: { [string]: string } }*/,
+) {
+  if (hmrAcceptCheckOne(bundle, id, depsByBundle)) {
+    return true;
+  }
+
+  // Traverse parents breadth first. All possible ancestries must accept the HMR update, or we'll reload.
+  let parents = getParents(module.bundle.root, id);
+  let accepted = false;
+  while (parents.length > 0) {
+    let v = parents.shift();
+    let a = hmrAcceptCheckOne(v[0], v[1], null);
+    if (a) {
+      // If this parent accepts, stop traversing upward, but still consider siblings.
+      accepted = true;
+    } else {
+      // Otherwise, queue the parents in the next level upward.
+      let p = getParents(module.bundle.root, v[1]);
+      if (p.length === 0) {
+        // If there are no parents, then we've reached an entry without accepting. Reload.
+        accepted = false;
+        break;
+      }
+      parents.push(...p);
+    }
+  }
+
+  return accepted;
+}
+
+function hmrAcceptCheckOne(
```

Comment:
So I guess the idea was that `hmrAcceptCheck` does the parent traversal and `hmrAcceptCheckOne` just checks if the single asset accepts, without the ancestry? But `hmrAcceptCheckOne` still calls `hmrAcceptCheck`, is that intentional?

Reply:
I think it should be ok because it only happens for the initial bundle, not when finding parents (because `depsByBundle` is null then).